### ISSUE TITLE
Question on DeiT training procedure

### ISSUE THREAD

**Issue by khvmaths** [lucidrains/vit-pytorch#103]
Hi all! I just wanna confirm one confusion.

1. Is it true that we could not get the training accuracy because we have a teacher to monitor that?
2. Is it possible to fine-tune the DeiT with distiller model? I mean, by removing the teacher network and fine-tune it.

Attached is my code to do the training, please verify if this is correct way of doing so. (I fine-tuned my teacher model on the dataset using FastAI).

a. Loading the fine-tuned teacher model.

```
from fastai.vision.all import *

loc = torch.load('v2/models/resnet50.pth')
body = create_body(resnet50)
head = create_head(2048, 10, None, ps=0.5, bn_final=False)
teacher = nn.Sequential(body, head)
teacher.load_state_dict(loc['model'])
```

b. Create DeiT model

```
v = DistillableViT(
    image_size = image_sz,
    patch_size = patch_sz,
    num_classes = 10,
    dim = 1024,
    depth = 6,
    heads = 8,
    mlp_dim = 2048,
    dropout = 0.1,
    emb_dropout = 0.1
)

model = DistillWrapper(
    student = v,
    teacher = teacher,
    temperature = 3,           # temperature of distillation
    alpha = 0.5,               # trade between main loss and distillation loss
    hard = False               # whether to use soft or hard distillation
).to(device)
```

c. Training [Do the criterion needed in this case?]

```
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=lr)
scheduler = StepLR(optimizer, step_size=1, gamma=gamma)

for epoch in range(epochs):
    epoch_loss = 0

    for data, label in tqdm(train_loader):
        data = data.to(device)
        label = label.to(device)
        loss = model(data, label)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        epoch_loss += loss / len(train_loader)

    with torch.no_grad():
        epoch_val_accuracy = 0
        epoch_val_loss = 0
        for data, label in valid_loader:
            data = data.to(device)
            label = label.to(device)

            val_output = v(data)
            val_loss = criterion(val_output, label)

            acc = (val_output.argmax(dim=1) == label).float().mean()
            epoch_val_accuracy += acc / len(valid_loader)
            epoch_val_loss += val_loss / len(valid_loader)

    print(
        f"Epoch : {epoch+1} - loss : {epoch_loss:.4f} - val_loss : {epoch_val_loss:.4f} - val_acc: {epoch_val_accuracy:.4f}\n"
    )
```

Thank you in advance!

**Comment by QueeneTam** [comment]
How to use dataparallel in distilling correctly?

**Comment by abhishekmann** [comment]
@hongvin Did you get the distillation procedure to work?

Is the way you mentioned right way of doing it?
